src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/VsTestConsoleRequestSender.cs

```diff
@@ -337,6 +470,60 @@ private void SendMessageAndListenAndReportTestCases(IEnumerable<string> sources,
             this.testPlatformEventSource.TranslationLayerDiscoveryStop();
         }
 
+        private async Task SendMessageAndListenAndReportTestCasesAsync(IEnumerable<string> sources, string runSettings, ITestDiscoveryEventsHandler eventHandler)
+        {
+            try
+            {
+                this.communicationManager.SendMessage(
+                            MessageType.StartDiscovery,
+                            new DiscoveryRequestPayload() { Sources = sources, RunSettings = runSettings },
+                            this.protocolVersion);
+                var isDiscoveryComplete = false;
+
+                // Cycle through the messages that the vstest.console sends.
+                // Currently each of the operations are not separate tasks since they should not each take much time.
+                // This is just a notification.
+                while (!isDiscoveryComplete)
```

Comment:
I imagine something like this...

```csharp
public Task DiscoverTestsAsync()
{
    var discoveryCompletionSource = new TaskCompletionSource<int>();

    this.communicationManager.MessageReceived += (sender, e) => 
    {
        switch(e.Type)
        {
            case MessageType.DiscoveryComplete:
                discoveryCompletionSource.SetResult(0);
                break;
        }
    };

    this.communicationManager.SendMessage(startDiscoveryMessage);

    await discoveryCompletionSource.Task;
}
```

And equivalently for DiscoverTests().

Unless you're planning on exposing an event-based interface publicly?